src/main/java/io/github/opencubicchunks/cubicchunks/cubicgen/customcubic/CustomGeneratorSettings.java

```diff
@@ -301,237 +360,198 @@ public static CustomGeneratorSettings fromVanilla(ChunkGeneratorSettings setting
-                        Blocks.DIAMOND_ORE.getDefaultState(),
-                        Blocks.EMERALD_ORE.getDefaultState(),
-                        Blocks.MONSTER_EGG.getDefaultState().withProperty(BlockSilverfish.VARIANT, BlockSilverfish.EnumType.STONE),
-                        Blocks.GOLD_ORE.getDefaultState()
-                };
-                Biome[][] standardBiomes = {
-                        null, // dirt
-                        null, // gravel
-                        null, // granite
-                        null, // diorite
-                        null, // andesite
-                        null, // coal
-                        null, // iron
-                        null, // gold
-                        null, // redstone
-                        null, // diamond
-                        {Biomes.EXTREME_HILLS, Biomes.EXTREME_HILLS_EDGE, Biomes.EXTREME_HILLS_WITH_TREES, Biomes.MUTATED_EXTREME_HILLS,
-                                Biomes.MUTATED_EXTREME_HILLS_WITH_TREES},//emerald
-                        {Biomes.EXTREME_HILLS, Biomes.EXTREME_HILLS_EDGE, Biomes.EXTREME_HILLS_WITH_TREES, Biomes.MUTATED_EXTREME_HILLS,
-                                Biomes.MUTATED_EXTREME_HILLS_WITH_TREES},//monster egg
-                        {Biomes.MESA, Biomes.MESA_CLEAR_ROCK, Biomes.MESA_ROCK, Biomes.MUTATED_MESA, Biomes.MUTATED_MESA_CLEAR_ROCK,
-                                Biomes.MUTATED_MESA_ROCK},//mesa gold
-                };
-                for (int i = 0; i < standard.length; i++) {
-                    JsonObject obj = convertStandardOre(gson, root, standard[i], standardBlockstates[i], standardBiomes[i]);
-                    if (obj != null) {
-                        standardOres.add(obj);
-                    }
-
-                }
-                JsonObject lapis = convertGaussianPeriodicOre(gson, root, "lapisLazuli", Blocks.LAPIS_ORE.getDefaultState(), null);
-                if (lapis != null) {
-                    periodicGaussianOres.add(lapis);
-                }
-                root.add("standardOres", standardOres);
-                root.add("periodicGaussianOres", periodicGaussianOres);
-                compound.setString("generatorOptions", gson.toJson(root));
-                return compound;
+    
+    public static String fixGeneratorOptionsIfNecessary(String generatorOptions) {
+        if (generatorOptions.isEmpty()) {
+            generatorOptions = new CustomGeneratorSettings().toJson(false);
+        }
+        JsonReader reader = new JsonReader(new StringReader(generatorOptions));
+        JsonObject root = new JsonParser().parse(reader).getAsJsonObject();
+        if(!root.has("version") ||root.get("version").getAsInt()<3) {
+            generatorOptions = fixGeneratorOptionsVersion0(generatorOptions);
+            generatorOptions = fixGeneratorOptionsVersion1(generatorOptions);
+            generatorOptions = fixGeneratorOptionsVersion3(generatorOptions);
```

Comment:
This is an awful way to do it. it may *work* (and even that isn't very obvious) but it definitely needs a rewrite.

Each of those functions was more or less ok at converting from previous to next version, but it became a pile of hacks.

Instead, it would be better to just implement it from the scratch, convert the original json to JsonObject, make a new JsonObject, and for each value in the v3 format write code to try to extract it from whatever the old format was.